### PR TITLE
ETL-4470: Implementation of http filer puler using Distcp-ng

### PR DESCRIPTION
**Pull request by ydai1124**
https://docs.google.com/document/d/1EmjQLmpvbFcBcFdCQSC46MkQ_EiowGDG8WLt729Zwdg/

### CODE REVIEW

**Comment by ydai1124** [comment]
@ibuenros Can you review?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java [hunk 1]
```diff
@@ -375,22 +454,107 @@ public static void serializeCopyEntity(State state, CopyEntity copyEntity) {
...
+      throws IOException {
     return CopyableDatasetMetadata.deserialize(state.getProp(SERIALIZED_COPYABLE_DATASET));
   }
 
+  /**
+   * Return the mapping from datasetUrn to {@link IncludeExcludeWatermark} based on previous workunitstates if filterByWatermark is enabled and watermarkGenerator is provided.
+   * If all {@link WorkUnitState}s are committed, watermark value will be the highest high watermark among all, and this value should be excluded for next time.
+   * Otherwise, it will be the lowest low watermark among all non-committed {@link WorkUnitState}s, and this value should included for next time.
+   */
+  public Map<String, IncludeExcludeWatermark> getPreivousWatermarkForFilterIfNecessary(boolean filterByWatermark,
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java]: typo: Preivous -> Previous

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java [hunk 2]
```diff
@@ -306,16 +349,49 @@ private boolean shouldStopGeneratingWorkUnits(ConcurrentBoundedWorkUnitList work
...
+      throws IOException {
+    if (!watermarkGenerator.isPresent()) {
+      return Optional.absent();
+    }
+    return watermarkGenerator.get().generateWatermarkIntervalForCopyEntity(copyEntity);
+  }
 
+  protected FileSystem getSourceFileSystem(State state)
+      throws IOException {
+    if (state.contains(CUSTOMIZED_FILESYSTEM)) {
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java]: Usually this should be done by Hadoop itself. When you provide a fs URI, Hadoop will use the scheme to try to find a FileSystem matching that scheme. To provide custom `FileSystem` implementations you can either use the configuration "fs.<scheme>.impl", or register the implementation as a service provider for `FileSystem` (creating a file `META-INF/service/org.apache.hadoop.fs.FileSystem` containing the name of the implementation).

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java [hunk 3]
```diff
@@ -246,13 +280,20 @@ public Void call() {
...
           Extract extract = new Extract(Extract.TableType.SNAPSHOT_ONLY, CopyConfiguration.COPY_PREFIX, extractId);
           List<WorkUnit> workUnitsForPartition = Lists.newArrayList();
           for (CopyEntity copyEntity : fileSet.getFiles()) {
 
             CopyableDatasetMetadata metadata = new CopyableDatasetMetadata(this.copyableDataset);
             CopyEntity.DatasetAndPartition datasetAndPartition = copyEntity.getDatasetAndPartition(metadata);
-
+            Optional<WatermarkInterval> watermarkIntervalOptional =
+                getCopyEntityWatermark(copyEntity, watermarkGenerator);
+            if (shouldSkipDueToWatermark(filterWuByWatermark,
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java]: Is there a way that the `Dataset` itself can use the watermark? This is doing a late filtering, but the dataset might be able to optimize this filtering, or use the watermark to do a faster listing.

**Comment by ibuenros** [comment]
Would it make sense to move all of the watermark functionality to a helper class? `CopySource` is getting pretty long.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/SubsetFilesCopyableDatasetFinder.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+
+/**
+ * {@link SubsetFilesCopyableDataset} finder that extends {@link ConfigurableGlobDatasetFinder}.
+ *
+ * It lists files on the root path, and uses a user defined pattern to identify {@link SubsetFilesCopyableDataset}s and
+ * their corresponding subset of files.
+ */
+@Slf4j
+public class SubsetFilesCopyableDatasetFinder extends ConfigurableGlobDatasetFinder<CopyableDataset> {
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/SubsetFilesCopyableDatasetFinder.java]: Can't this functionality be achieved with `RecursiveCopyableDataset` itself? If I understand correctly, you can just use the identifier pattern to assign files to a file set, and add a filter.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/SubsetFilesCopyableDatasetFinder.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  public static final String IDENTIFIER_PATTERN = CopyConfiguration.COPY_PREFIX + ".subsetFilesDatasetIdentifer";
+  public static final String DEFAULT_IDENTIFIER_PATTERN = "([\\w_\\d]*\\.[\\w_\\d]*)\\..*(gz|meta)";
+
+  public static final String SUBSETFILES_REGEX_FILTER = CopyConfiguration.COPY_PREFIX + ".subsetFilesRegexFilter";
+  public static final String DEFAULT_SUBSETFILES_REGEX_FILTER = ".*(gz|meta)";
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/SubsetFilesCopyableDatasetFinder.java]: I don't think this is a sensible default for open source.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/watermark/StringBasedCopyEntityWatermarkGenerator.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import gobblin.data.management.copy.CopyEntity;
+import gobblin.data.management.copy.CopyableFile;
+import gobblin.source.extractor.ComparableWatermark;
+import gobblin.source.extractor.WatermarkInterval;
+
+
+/**
+ * Implementation of {@link CopyEntityWatermarkGenerator} that uses {@link StringWatermark}.
+ */
+public class StringBasedCopyEntityWatermarkGenerator implements CopyEntityWatermarkGenerator {
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/watermark/StringBasedCopyEntityWatermarkGenerator.java]: Can you rename this to something more descriptive? Like `FullPathCopyEntityWatermarkGenerator`?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/watermark/StringBasedCopyEntityWatermarkGenerator.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import com.google.common.base.Optional;
+
+import gobblin.data.management.copy.CopyEntity;
+import gobblin.data.management.copy.CopyableFile;
+import gobblin.source.extractor.ComparableWatermark;
+import gobblin.source.extractor.WatermarkInterval;
+
+
+/**
+ * Implementation of {@link CopyEntityWatermarkGenerator} that uses {@link StringWatermark}.
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/watermark/StringBasedCopyEntityWatermarkGenerator.java]: Please explain what string will be returned.

File: gobblin-utility/src/main/java/gobblin/util/io/HttpInputStream.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+
+import org.apache.hadoop.fs.FSDataInputStream;
+
+
+/**
+ * An extension of {@link FSDataInputStream} that takes a {@link HttpURLConnection} and closes the connection in {@link #close()} method.
+ */
+public class HttpInputStream extends FSDataInputStream {
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/io/HttpInputStream.java]: This could be more generic, taking any `Closeable` or even a set of `Closeables` that will al be closed at the same time as the `InputStream`.

File: gobblin-core/build.gradle [hunk 9]
```diff
@@ -36,6 +36,8 @@ dependencies {
   compile externalDependency.commonsIo
   compile externalDependency.hiveExec
   compile externalDependency.hiveSerDe
+  compile externalDependency.htmlparser
+  compile externalDependency.htmllexer
```
Comment (ibuenros) [inline gobblin-core/build.gradle]: What are these being used for?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/SubsetFilesCopyableDatasetFinder.java (hunk 4 above)
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/SubsetFilesCopyableDatasetFinder.java]: It is not recursively finding datasets. The non-recursive listing phase is done in `datasetFinder` rather than `dataset` for performance purpose. Limited logic in `RecursiveCopyableDataset` can be reused here.

**Comment by ydai1124** [comment]
@ibuenros Updated.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/SubsetFilesCopyableDatasetFinder.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    }
+    return datasets;
+  }
+
+  public CopyableDataset datasetAndPathWithIdentifier(Path path, String identifier)
+      throws IOException {
+    List<Object> args =
+        Lists.newArrayList(fs, path, props, identifier, idToFileStatuses.get(identifier), eventSubmitter, state);
+    try {
+      return GobblinConstructorUtils.invokeLongestConstructor(SubsetFilesCopyableDataset.class, args.toArray());
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/SubsetFilesCopyableDatasetFinder.java]: You can just call `GobblinConstructorUtils.invokeLongestConstructor(SubsetFilesCopyableDataset.class, fs, path, props, identifier, idToFileStatuses.get(identifier), eventSubmitter, state)`

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/SubsetFilesCopyableDatasetFinder.java (hunk 4 above)
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/SubsetFilesCopyableDatasetFinder.java]: There is a lot of repeated logic between this and the recursive copyable dataset, and features that that one would benefit from (like using a pattern to define filesets). Can you merge them as much as possible?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/WatermarkBasedCopyableFileFilter.java [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+/**
+ * Implementation of {@link CopyableFileFilter} that filters {@link CopyableFile}s based on previous watermark information.
+ */
+@Slf4j
+public class WatermarkBasedCopyableFileFilter implements CopyableFileFilter {
+  /**
+   * Only create work units that have higher expected high watermark than the previous actual watermark.
+   */
+  public static final String FILTER_BY_WATERMARK = CopyConfiguration.COPY_PREFIX + ".filterWorkUnitsByWatermark";
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/WatermarkBasedCopyableFileFilter.java]: Why is this needed? Is this filter always instantiated or is it user specified? If the latter, wouldn't the user specifying this filter imply that they want the filtering?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/WatermarkBasedCopyableFileFilter.java [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    }
+  }
+
+  /**
+   * Return the mapping from datasetUrn to {@link IncludeExcludeWatermark} based on previous workunitstates if filterByWatermark is enabled and watermarkGenerator is provided.
+   * If all {@link WorkUnitState}s are committed, watermark value will be the highest high watermark among all, and this value should be excluded for next time.
+   * Otherwise, it will be the lowest low watermark among all non-committed {@link WorkUnitState}s, and this value should included for next time.
+   */
+  @VisibleForTesting
+  protected Map<String, IncludeExcludeWatermark> getPreviousWatermarkForFilterIfNecessary(boolean filterByWatermark,
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/WatermarkBasedCopyableFileFilter.java]: It seems that a new Map is created for each dataset, even though all maps are the same. Can this be instantiated once and reused?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/watermark/CopyEntityWatermarkGenerator.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+
+/**
+ * Watermark generator for {@link CopyEntity}.
+ */
+public interface CopyEntityWatermarkGenerator {
+  /**
+   * Generate optional {@link WatermarkInterval} for a given {@link CopyEntity}.
+   */
+  public Optional<WatermarkInterval> generateWatermarkIntervalForCopyEntity(CopyEntity copyEntity) throws IOException;
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/watermark/CopyEntityWatermarkGenerator.java]: Above you only use watermarks for `CopyableFile`, is there a point to allowing watermarks for any `CopyEntity`?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/WatermarkBasedCopyableFileFilter.java (hunk 12 above)
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/WatermarkBasedCopyableFileFilter.java]: It is actually only instantiated once. `STATE_TO_PREV_WATERMARK_MAP_CAHCE` stores the mapping from `SourceState` reference to `Map`.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/watermark/CopyEntityWatermarkGenerator.java (hunk 13 above)
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/watermark/CopyEntityWatermarkGenerator.java]: Not for my case. I cannot really come up a good use case for other `CopyEntity`. Will remove that.

**Comment by ydai1124** [comment]
@ibuenros  Updated.

**Comment by ibuenros** [comment]
LGTM
